fw/access_log.c

```diff
@@ -0,0 +1,246 @@
+/**
+ *		Tempesta FW
+ *
+ * Copyright (C) 2014 NatSys Lab. (info@natsys-lab.com).
+ * Copyright (C) 2015-2022 Tempesta Technologies, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+#include "access_log.h"
+#include "connection.h"
+#include "server.h"
+#include "http.h"
+#include "lib/common.h"
+
+static bool access_log_enabled = false;
+
+/* Use small buffer because printk won't display strings longer that ~1000 bytes */
+#define ACCESS_LOG_BUF_SIZE 990
+static DEFINE_PER_CPU_ALIGNED(char[ACCESS_LOG_BUF_SIZE], access_log_buf);
+
+/** Build string consists of chunks that belong to the header value.
+ * If header value is empty, then it returns "-" to be nginx-like.
+ *
+ * @param http_version is TFW_HTTP_VER_xx enum
+ * @param line is the request line from TfwHttpReq::h_tbl */
+static TfwStr
+get_http_header_value(char http_version, TfwStr *line)
+{
+	TfwStr result;
+	size_t len = line->len;
+	TfwStr *chunk = line->chunks, *end = chunk + line->nchunks;
+	static TfwStr empty_hdr = TFW_STR_STRING("-");
+
+	if (http_version >= TFW_HTTP_VER_09 && http_version <= TFW_HTTP_VER_11) {
+
+		/* Should never get plain string from parser */
+		if (TFW_STR_PLAIN(line))
+			return TFW_STR_EMPTY(line) ? empty_hdr : *line;
+
+		/* Skip over until ':' marker */
+		for (; chunk < end; chunk++) {
+			len -= chunk->len;
+			if (chunk->len == 1 && *chunk->data == ':')
+				break;
+		}
+		if (chunk == end)
+			return empty_hdr;
+		chunk++;
+
+		/* Skip possible whitespace blocks */
+		while (chunk < end && (chunk->flags & TFW_STR_OWS) != 0) {
+			len -= chunk->len;
+			chunk++;
+		}
+	} else {
+		/* skip over chunks until HDR_VALUE one */
+		while (chunk < end && (chunk->flags & TFW_STR_HDR_VALUE) == 0) {
+			len -= chunk->len;
+			chunk++;
+		}
```

Comment:
I guess we should do something different for `__TFW_HTTP_VER_INVALID`, probably not try to print any value. When this can happen? If it's unexpected, then it makes sense to do `WARN_ON_ONCE()` for this case.

Reply:
Such things only occur on malformed requests (i.e. when there was an error detected by parser before or in that field). I've decided  to go with `'-'` replacement to make it easier to understand what fields were malformed — this could be useful for debugging I think.
Ideally it would be nice to get original unmodified request line, but that doesn't seem to be easily possible.